Endoflife api/v1

marcwrobel opened the pull request:
This is a major rework of the API that fixes a lots of issues, but introduces a lot of breaking changes. See https://github.com/endoflife-date/endoflife.date/blob/2062-expose-product-info-api/CHANGELOG_API.md for more information.

Note that I thought of disabling API generation in development (using JEKYLL_ENV like the Jekyll Feed plugin - see https://github.com/jekyll/jekyll-feed/blob/master/lib/jekyll-feed/generator.rb#L145), but I finally reverted it. It does not work well with Netlify preview, and generate production URL (i.e. https://endoflife.date/ URLs) in development which makes testing more difficult.

marcwrobel committed:
First version

marcwrobel committed:
Add API v1 (#2595, #2425, #2331, #2066, #2062, #1762, #759, #394, #2530)

This is a major rework of the API with a lot of breaking changes. See CHANGELOG_API.md for more information.

Note that we thought of disabling API generation in development (using JEKYLL_ENV like the Jekyll Feed plugin - see https://github.com/jekyll/jekyll-feed/blob/master/lib/jekyll-feed/generator.rb#L145), but it was finally reverted. It does not work well with Netlify preview, and generate production URL (i.e. https://endoflife.date URLs) in development which makes it difficult to use.

captn3m0 commented on _plugins/generate-api-v1.rb:
```diff
@@ -0,0 +1,92 @@
+        has_releases = (page.data.has_key? 'releases')
+        is_product_page = (is_markdown and has_permalink and has_releases)
+
+        if is_product_page
```
A simpler check would be `page['layout'] == 'product'`

marcwrobel replied:
You are right, don't know why I made something this complicated.

hugovk commented on CHANGELOG_API.md:
```diff
@@ -0,0 +1,54 @@
+## API v0 alpha
```
?

```suggestion
## API v1 alpha
```

And further on below it says beta:

```suggestion
## API v1 beta
```

marcwrobel replied:
I choose to name the current API `v0`, and because it was an alpha version I added `alpha` to the name. The new API is labelled `v1`. I think I will drop the `alpha` and `beta` mention because it is confusing.

The order in the changelog may be confusion too (oldest first), will change it.

marcwrobel replied:
Done.

hugovk commented on CHANGELOG_API.md:
```diff
@@ -0,0 +1,54 @@
+## API v0 alpha
+
+On 2023-03-02 endpoints were:
```
```suggestion
On 2023-03-02 the v0 endpoints were:
```

marcwrobel replied:
Applied, thanks.

hugovk commented on CHANGELOG_API.md:
```diff
@@ -0,0 +1,54 @@
+On 2023-03-02 endpoints were:
+
+- "All products" (`/api/all.json`) : Return a list of all products.
+- "Product" (`/api/{product}.json`) : Get EoL dates of all cycles of a given product.
```
```suggestion
- "Product" (`/api/{product}.json`) : Get EoL dates of all cycles for a given product.
```

marcwrobel replied:
Fixed.

hugovk commented on CHANGELOG_API.md:
```diff
@@ -0,0 +1,54 @@
+
+### Summary
+
+API v1 is a major rework of the API v0 with a lot of breaking changes. Compared the API v0, API v1:
```
```suggestion
API v1 is a major rework of the API v0 with a lot of breaking changes. Compared to the API v0, API v1:
```

marcwrobel replied:
Fixed.

hugovk commented on CHANGELOG_API.md:
```diff
@@ -0,0 +1,54 @@
+- is generated using a Jekyll Generator (see https://jekyllrb.com/docs/plugins/generators/),
+- is versioned using the `api/v1` prefix, making it easier to implement non-backward-compatible
+  changes,
+- feels more "Restful".
```
Reorder. The Jekyll implementation details are less interesting to end users than the versioning:

```suggestion
- is versioned using the `api/v1` prefix, making it easier to implement non-backward-compatible
  changes,
- feels more "RESTful".
- is generated using a Jekyll Generator (see https://jekyllrb.com/docs/plugins/generators/),
```

marcwrobel replied:
Done.

hugovk commented on CHANGELOG_API.md:
```diff
@@ -0,0 +1,54 @@
+API v1 resolves : #394, #759, #905, #2062, #2066, #2078, #2160, #2331, #2431, #2595. It also reverts
+#2425 due to incompatibilities in redirect rules.
+
+The API v0 is still generated to give time to users to migrate to API v1.
```
I guess it may depend on how quickly users migrate, but can we give a rough idea of how long v0 will still be available? That will help people plan migration.

marcwrobel replied:
For the time being v0 does not require much maintenance. And I do not see it becoming a problem soon because we can emulate the old frontmatter format using `product-data-enricher.rb`. So maybe we could say 1 year from the API v1 release date ? Would that be enough ?

@endoflife-date/everyone, any opinion regarding this subject ?

hugovk replied:
I think one year should be plenty of time.

captn3m0 replied:
1 year sounds good. We could re-enable Netlify Analytics at the end of the year and evaluate if we need to keep is running further. If there's not much maintenance required, I don't mind keeping it even further.

adriens replied:
1 year is :ok_hand: 
And when I see all the data that will available, it will motivate to put effort on migrations around it :100:

hugovk commented on CHANGELOG_API.md:
```diff
@@ -0,0 +1,54 @@
+### Changes in 404 error responses
+
+404 error JSON responses are not returned anymore. #2425 has been reverted because it conflicted
+with the rule that rewrite the paths to add `/index.json` to all requests, which is also a global
```
```suggestion
with the rule that rewrites the paths to add `/index.json` to all requests, which is also a global
```

marcwrobel replied:
Fixed.

hugovk commented on HACKING.md:
```diff
@@ -105,7 +105,7 @@ The API is just JSON files generated in the `api` directory by `_plugins/create-
 ### API Documentation
 
-The API Documentation is available at <https://endoflife.date/docs/api> and is generated from an OpenAPI Specification file located at `_data/openapi.yml`. The documentation is rendered [Stoplight Elements](https://meta.stoplight.io/docs/elements/ZG9jOjMyNjU4OTY0-introduction-to-elements).
+The API Documentation is available at <https://endoflife.date/docs/api/v1/> and is generated from an OpenAPI Specification file located at `_data/openapi.yml`. The documentation is rendered [Stoplight Elements](https://meta.stoplight.io/docs/elements/ZG9jOjMyNjU4OTY0-introduction-to-elements).
```
```suggestion
The API Documentation is available at <https://endoflife.date/docs/api/v1/> and is generated from an OpenAPI Specification file located at `_data/openapi.yml`. The documentation is rendered by [Stoplight Elements](https://meta.stoplight.io/docs/elements/ZG9jOjMyNjU4OTY0-introduction-to-elements).
```

marcwrobel replied:
In fact the entire paragraph must be updated to mention both the old and new API documentation. Done.

hugovk commented on _redirects:
```diff
@@ -9,7 +9,14 @@
 ---
-{%- for page in site.pages -%}
+# Rewrite for /api/v1/ to keep URLs clean.
+# All API responses are located in an index.json and must be accessible without the file name, such as :
```
```suggestion
# All API responses are located in an index.json and must be accessible without the file name, such as:
```

marcwrobel replied:
Fixed.

hugovk commented on _redirects:
```diff
@@ -9,7 +9,14 @@
+# All API responses are located in an index.json and must be accessible without the file name, such as :
+# - /api/v1/index.json -> /api/v1/
+# - /api/v1/products/almalinux/index.json -> /api/v1/products/almalinux/
+# This is using shadowing : https://docs.netlify.com/routing/redirects/rewrites-proxies/#shadowing.
```
```suggestion
# This uses shadowing: https://docs.netlify.com/routing/redirects/rewrites-proxies/#shadowing.
```

marcwrobel replied:
Fixed.

hugovk reviewed (commented):
This changelog is good, shows clearly what will change to help people migrate 👍

adriens reviewed (approved):
LGTM and having some cool ideas to share the data for datascientists. :smile_cat:

usta reviewed (approved):
LGTM

BiNZGi reviewed (approved):
I tried several endpoints/combinations at https://deploy-preview-2080--endoflife-date.netlify.app/docs/api/v1/ and all LGTM. I vote for merging this in current shape :+1:

jamietanna commented on api_v1/openapi.yml:
```diff
@@ -0,0 +1,531 @@
+          description: Release date of the cycle.
+          example: "2022-04-21"
+        isLts:
+          ype: boolean
```
```suggestion
          type: boolean
```

marcwrobel replied:
Fixed, thanks @jamietanna !

jamietanna commented on api_v1/openapi.yml:
```diff
@@ -0,0 +1,531 @@
+      type: object
+      properties:
+        name:
+          type: type
```
This should be maybe:

```suggestion
          type: string
```

marcwrobel replied:
Fixed, thanks @jamietanna !

marcwrobel committed:
Reviews

- improve openapi schema
- upgrade to openapi 3.1 and apply https://www.openapis.org/blog/2021/02/16/migrating-from-openapi-3-0-to-3-1-0
- upgrade to swagger-ui v5
- rename cycle -> release
- rename release.date -> release.releaseDate
- make some fields optional (eoas, eoes, discontinued)

Co-authored-by: Jamie Tanna <jamie@jamietanna.co.uk>

jamietanna reviewed (commented):
On a general point, this is looking really good! I appreciate that we're now able to see Package URL (pURLs) if present, which will make some of the integrations I do with `endoflife.date` much easier

It'd be awesome if there was an easier way to "look up" whether a pURL is understood by `endoflife.date` but I realise that this is a _static_ API and so making this make sense could be more difficult

I wonder if maybe an endpoint that lists the known pURLs and what products they correspond to?

(happy for this to be a follow-up)

jamietanna commented on api_v1/openapi.yml:
I notice that we don't have a set of `required` properties on a few of the objects, i.e. `ProductReleaseResponse` - is that expected?

marcwrobel replied:
Nope, that's a mistake. Will soon have a look at it.

marcwrobel replied:
@jamietanna this should be fixed now.

I improved the API doc and schema to, hopefully, make things clearer (all changes are in https://github.com/endoflife-date/endoflife.date/pull/2080/commits/f5c6ce1a62cb02334cb4a442d86c48e624b93490). Let me know what you think when you have some time.

jamietanna replied:
Looking good so far! I tested this with [Wiretap](https://github.com/pb33f/wiretap) as a way to validate that the OpenAPI spec matches our implementation, and noticed a few things:

- `/`
  - Is missing a `total`
- `/products/ruby`
- `/products/java`
  - isEoes: `null`, wanted `boolean`
- `/products/nodejs/releases/22`
  - isEoes: `null`, wanted `boolean`

It looks like most of the products I've tested with `java`, `ruby`, `go` all have the same:

- Is missing `/properties/result/properties/identifiers/items/required`: `name`
  - i.e. it `identifiers[0].name` is missing

<details>



```json
{
  "generated_at": "2025-04-26T11:22:45+00:00",
  "last_modified": "2025-04-02T13:03:42+00:00",
  "result": {
    "aliases": [
      "golang"
    ],
    "category": "lang",
    "identifiers": [
      {
        "id": "go",
        "type": "repology"
      },
      {
        "id": "pkg:generic/go",
        "type": "purl"
      },
      {
        "id": "pkg:docker/library/golang",
        "type": "purl"
      },
      {
        "id": "pkg:docker/circleci/golang",
        "type": "purl"
      },
      {
        "id": "pkg:docker/cimg/go",
        "type": "purl"
      },
      {
        "id": "pkg:docker/bitnami/golang",
        "type": "purl"
      }
    ]
  },
  "schema_version": "1.0.0"
}
```

</details>

<details>

<summary>Testing instructions</summary>

```sh
npx @pb33f/wiretap@latest -s https://deploy-preview-2080--endoflife-date.netlify.app/docs/api/v1/openapi.yml -u https://deploy-preview-2080--endoflife-date.netlify.app/api/v1
```

- Open up the local web UI (http://localhost:9091 for me)
- Start sending API requests i.e. `curl -i http://localhost:9090/products/ruby`
- In the Web UI, notice that the validation fails for some requests

</details>

marcwrobel replied:
Thanks a lot @jamietanna, except the missing `total` those were errors in the spec. It's now fixed. I will do some more checks this week with `wiretap` to see if there  are some more errors.

I think it would be nice to have a GHA workflow validating the spec as well as the generated JSON files. Do you know a good CLI supporting OpenAPI 3.1 that I could use ?

marcwrobel committed:
ci
